File: go/mutation_test.go
package servirtium_test

import (
	"strings"
	"testing"

	servirtium "github.com/servirtium/servirtium-go"
)

// Record-mode breadth: redaction, header removal, notes, drift detection,
// and (critically) that per-handle mutation state does not leak from one
// fixture to the next (each server owns its own handle/state).

func TestRecordRedactsResponseBody(t *testing.T) {
	up := newFakeUpstream()
	defer up.close()
	up.responseBody = "value=secret-token"
	tape := tempTape(t, "rec.md")

	rec, err := servirtium.Record(tape, up.baseURL()).
		Redact(servirtium.ResponseBody, "secret-token", "REDACTED").
		Port(0).
		Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, rec.BaseURL(), "/x")
	if err := rec.Close(); err != nil {
		t.Fatal(err)
	}

	tapeText := readTape(t, tape)
	if !strings.Contains(tapeText, "REDACTED") {
		t.Fatalf("expected tape to contain REDACTED:\n%s", tapeText)
	}
	if strings.Contains(tapeText, "secret-token") {
		t.Fatalf("tape must not contain the secret:\n%s", tapeText)
	}
}

func TestRecordAttachesNote(t *testing.T) {
	up := newFakeUpstream()
	defer up.close()
	tape := tempTape(t, "rec.md")

	rec, err := servirtium.Record(tape, up.baseURL()).
		Note("Why this exists", "documents the call").
		Port(0).
		Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, rec.BaseURL(), "/x")
	if err := rec.Close(); err != nil {
		t.Fatal(err)
	}

	tapeText := readTape(t, tape)
	if !strings.Contains(tapeText, "## [Note] Why this exists:") {
		t.Fatalf("expected note heading on tape:\n%s", tapeText)
	}
}

func TestRecordRemovesNamedResponseHeader(t *testing.T) {
	up := newFakeUpstream()
	defer up.close()
	up.extraHeaders["X-Trace-Id"] = "abc123"
	tape1 := tempTape(t, "rec1.md")
	tape2 := tempTape(t, "rec2.md")

	// Phase 1: without removal, the header is captured on the tape.
	rec, err := servirtium.Record(tape1, up.baseURL()).Port(0).Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, rec.BaseURL(), "/x")
	if err := rec.Close(); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(readTape(t, tape1), "X-Trace-Id") {
		t.Fatal("phase 1: expected X-Trace-Id on the tape")
	}

	// Phase 2: with removal, it's gone.
	rec2, err := servirtium.Record(tape2, up.baseURL()).
		RemoveHeader(servirtium.ResponseHeaders, "X-Trace-Id").
		Port(0).
		Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, rec2.BaseURL(), "/x")
	if err := rec2.Close(); err != nil {
		t.Fatal(err)
	}
	if strings.Contains(readTape(t, tape2), "X-Trace-Id") {
		t.Fatalf("phase 2: X-Trace-Id should have been removed:\n%s", readTape(t, tape2))
	}
}

func TestMutationStateDoesNotLeakBetweenFixtures(t *testing.T) {
	up := newFakeUpstream()
	defer up.close()
	tape1 := tempTape(t, "rec1.md")
	tape2 := tempTape(t, "rec2.md")

	// Fixture A registers a redaction for "leak".
	up.responseBody = "leak"
	a, err := servirtium.Record(tape1, up.baseURL()).
		Redact(servirtium.ResponseBody, "leak", "SCRUBBED").
		Port(0).
		Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, a.BaseURL(), "/x")
	if err := a.Close(); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(readTape(t, tape1), "SCRUBBED") {
		t.Fatal("fixture A: expected SCRUBBED")
	}

	// Fixture B registers NO redaction; A's must not leak in.
	b, err := servirtium.Record(tape2, up.baseURL()).Port(0).Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, b.BaseURL(), "/x")
	if err := b.Close(); err != nil {
		t.Fatal(err)
	}
	tape2Text := readTape(t, tape2)
	if !strings.Contains(tape2Text, "leak") {
		t.Fatalf("fixture B: expected the un-redacted body:\n%s", tape2Text)
	}
	if strings.Contains(tape2Text, "SCRUBBED") {
		t.Fatalf("fixture B: A's redaction leaked:\n%s", tape2Text)
	}
}

func TestFailIfChangedReturnsErrorOnDrift(t *testing.T) {
	up := newFakeUpstream()
	defer up.close()
	tape := tempTape(t, "rec.md")

	// First record creates the tape — no drift, no error.
	up.responseBody = "v1"
	first, err := servirtium.Record(tape, up.baseURL()).FailIfChanged().Port(0).Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, first.BaseURL(), "/x")
	if err := first.Close(); err != nil {
		t.Fatalf("first record should not drift: %v", err)
	}

	// Re-record with a changed upstream — Close must return an error, while
	// still writing the new tape for git diff.
	up.responseBody = "v2-changed"
	second, err := servirtium.Record(tape, up.baseURL()).FailIfChanged().Port(0).Start()
	if err != nil {
		t.Fatal(err)
	}
	get(t, second.BaseURL(), "/x")
	if err := second.Close(); err == nil {
		t.Fatal("expected a drift error from Close, got nil")
	}
	if !strings.Contains(readTape(t, tape), "v2-changed") {
		t.Fatal("drift: the new tape should still be written for git diff")
	}
}

File: zig/src/mutation_test.zig
//! Record-mode breadth, ported from the Go binding's `mutation_test.go`:
//! response-body redaction, notes, named-header removal, that per-handle
//! mutation state does not leak between fixtures, and fail-if-changed drift.

const std = @import("std");
const servirtium = @import("servirtium.zig");
const testutil = @import("testutil.zig");

const Field = servirtium.Field;

fn tapePathZ(allocator: std.mem.Allocator, tmp: *std.testing.TmpDir, name: []const u8) ![:0]u8 {
    const rel = try std.fs.path.join(allocator, &.{ ".zig-cache", "tmp", &tmp.sub_path, name });
    defer allocator.free(rel);
    return testutil.cwdJoinZ(allocator, rel);
}

fn readTape(allocator: std.mem.Allocator, path: [:0]const u8) ![]u8 {
    return std.Io.Dir.cwd().readFileAlloc(std.testing.io, path, allocator, .unlimited);
}

/// GET base+"/x" against a running server; discards the response.
fn poke(allocator: std.mem.Allocator, srv: *servirtium.Server) !void {
    const base = try srv.baseUrl();
    defer allocator.free(base);
    const url = try std.fmt.allocPrint(allocator, "{s}/x", .{base});
    defer allocator.free(url);
    var res = try testutil.curl(allocator, std.testing.io, .{ .url = url });
    res.deinit(allocator);
}

test "record redacts the response body before it lands on the tape" {
    const allocator = std.testing.allocator;

    var up = try testutil.FakeUpstream.init(allocator);
    defer up.deinit();
    up.response_body = "value=secret-token";
    const upstream = try up.baseUrl(allocator);
    defer allocator.free(upstream);

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const tape = try tapePathZ(allocator, &tmp, "rec.md");
    defer allocator.free(tape);

    var rec = try servirtium.Record.init(allocator, tape, upstream)
        .redact(Field.response_body, "secret-token", "REDACTED")
        .port(0)
        .start();
    try poke(allocator, &rec);
    try rec.close();

    const text = try readTape(allocator, tape);
    defer allocator.free(text);
    try std.testing.expect(std.mem.indexOf(u8, text, "REDACTED") != null);
    try std.testing.expect(std.mem.indexOf(u8, text, "secret-token") == null);
}

test "record attaches a note" {
    const allocator = std.testing.allocator;

    var up = try testutil.FakeUpstream.init(allocator);
    defer up.deinit();
    const upstream = try up.baseUrl(allocator);
    defer allocator.free(upstream);

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const tape = try tapePathZ(allocator, &tmp, "rec.md");
    defer allocator.free(tape);

    var rec = try servirtium.Record.init(allocator, tape, upstream)
        .note("Why this exists", "documents the call")
        .port(0)
        .start();
    try poke(allocator, &rec);
    try rec.close();

    const text = try readTape(allocator, tape);
    defer allocator.free(text);
    try std.testing.expect(std.mem.indexOf(u8, text, "## [Note] Why this exists:") != null);
}

test "record removes a named response header" {
    const allocator = std.testing.allocator;

    var up = try testutil.FakeUpstream.init(allocator);
    defer up.deinit();
    up.extra_header_name = "X-Trace-Id";
    up.extra_header_value = "abc123";
    const upstream = try up.baseUrl(allocator);
    defer allocator.free(upstream);

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const tape1 = try tapePathZ(allocator, &tmp, "rec1.md");
    defer allocator.free(tape1);
    const tape2 = try tapePathZ(allocator, &tmp, "rec2.md");
    defer allocator.free(tape2);

    // Phase 1: without removal, the header is captured on the tape.
    {
        var rec = try servirtium.Record.init(allocator, tape1, upstream).port(0).start();
        try poke(allocator, &rec);
        try rec.close();
        const text = try readTape(allocator, tape1);
        defer allocator.free(text);
        try std.testing.expect(std.mem.indexOf(u8, text, "X-Trace-Id") != null);
    }

    // Phase 2: with removal, it's gone.
    {
        var rec = try servirtium.Record.init(allocator, tape2, upstream)
            .removeHeader(Field.response_headers, "X-Trace-Id")
            .port(0)
            .start();
        try poke(allocator, &rec);
        try rec.close();
        const text = try readTape(allocator, tape2);
        defer allocator.free(text);
        try std.testing.expect(std.mem.indexOf(u8, text, "X-Trace-Id") == null);
    }
}

test "mutation state does not leak between fixtures" {
    const allocator = std.testing.allocator;

    var up = try testutil.FakeUpstream.init(allocator);
    defer up.deinit();
    up.response_body = "leak";
    const upstream = try up.baseUrl(allocator);
    defer allocator.free(upstream);

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const tape1 = try tapePathZ(allocator, &tmp, "rec1.md");
    defer allocator.free(tape1);
    const tape2 = try tapePathZ(allocator, &tmp, "rec2.md");
    defer allocator.free(tape2);

    // Fixture A registers a redaction for "leak".
    {
        var a = try servirtium.Record.init(allocator, tape1, upstream)
            .redact(Field.response_body, "leak", "SCRUBBED")
            .port(0)
            .start();
        try poke(allocator, &a);
        try a.close();
        const text = try readTape(allocator, tape1);
        defer allocator.free(text);
        try std.testing.expect(std.mem.indexOf(u8, text, "SCRUBBED") != null);
    }

    // Fixture B registers NO redaction; A's must not leak in.
    {
        var b = try servirtium.Record.init(allocator, tape2, upstream).port(0).start();
        try poke(allocator, &b);
        try b.close();
        const text = try readTape(allocator, tape2);
        defer allocator.free(text);
        try std.testing.expect(std.mem.indexOf(u8, text, "leak") != null);
        try std.testing.expect(std.mem.indexOf(u8, text, "SCRUBBED") == null);
    }
}

test "fail-if-changed returns an error on drift" {
    const allocator = std.testing.allocator;

    var up = try testutil.FakeUpstream.init(allocator);
    defer up.deinit();
    const upstream = try up.baseUrl(allocator);
    defer allocator.free(upstream);

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const tape = try tapePathZ(allocator, &tmp, "rec.md");
    defer allocator.free(tape);

    // First record creates the tape — no drift, no error.
    up.response_body = "v1";
    {
        var first = try servirtium.Record.init(allocator, tape, upstream).failIfChanged().port(0).start();
        try poke(allocator, &first);
        try first.close(); // should not drift
    }

    // Re-record with a changed upstream — close must return an error, while
    // still writing the new tape for git diff.
    up.response_body = "v2-changed";
    {
        var second = try servirtium.Record.init(allocator, tape, upstream).failIfChanged().port(0).start();
        try poke(allocator, &second);
        try std.testing.expectError(servirtium.Error.VcrError, second.close());
    }

    const text = try readTape(allocator, tape);
    defer allocator.free(text);
    try std.testing.expect(std.mem.indexOf(u8, text, "v2-changed") != null);
}
